cpp/src/arrow/compute/kernels/isin.cc

```diff
@@ -0,0 +1,282 @@
+                 std::shared_ptr<ArrayData>& out) override {
+    bool_builder_.Reset();
+    bool_builder_.Reserve(left.length);
+
+    if (left.null_count != 0 && right.null_count != 0) {
+      for (int64_t i = 0; i < left.length; ++i) {
+        bool_builder_.UnsafeAppend(true);
+      }
+    } else if (left.null_count != 0 && right.null_count == 0) {
+      for (int64_t i = 0; i < left.length; ++i) {
+        bool_builder_.UnsafeAppend(false);
+      }
+    }
+    RETURN_NOT_OK(bool_builder_.FinishInternal(&out));
+    return Status::OK();
+  }
+
+ private:
+  BooleanBuilder bool_builder_;
+};
+
+// ----------------------------------------------------------------------
+// Kernel wrapper for generic hash table kernels
+
+template <typename Type, typename Enable = void>
+struct HashBinaryKernelTraits {};
+
+template <typename Type>
+struct HashBinaryKernelTraits<Type, enable_if_null<Type>> {
+  using HashBinaryKernelImpl = NullIsInKernel;
+};
+
+template <typename Type>
+struct HashBinaryKernelTraits<Type, enable_if_has_c_type<Type>> {
+  using HashBinaryKernelImpl = IsInKernel<Type, typename Type::c_type>;
+};
+
+template <typename Type>
+struct HashBinaryKernelTraits<Type, enable_if_boolean<Type>> {
+  using HashBinaryKernelImpl = IsInKernel<Type, bool>;
+};
+
+template <typename Type>
+struct HashBinaryKernelTraits<Type, enable_if_binary<Type>> {
+  using HashBinaryKernelImpl = IsInKernel<Type, util::string_view>;
+};
+
+template <typename Type>
+struct HashBinaryKernelTraits<Type, enable_if_fixed_size_binary<Type>> {
+  using HashBinaryKernelImpl = IsInKernel<Type, util::string_view>;
+};
+
+Status GetIsInKernel(FunctionContext* ctx, const std::shared_ptr<DataType>& type,
+                     std::unique_ptr<HashBinaryKernelImpl>* out) {
+  std::unique_ptr<HashBinaryKernelImpl> kernel;
+
+#define ISIN_CASE(InType)                                                           \
+  case InType::type_id:                                                             \
+    kernel.reset(new typename HashBinaryKernelTraits<InType>::HashBinaryKernelImpl( \
+        type, ctx->memory_pool()));                                                 \
+    break
+
+  switch (type->id()) {
+    ISIN_CASE(NullType);
+    ISIN_CASE(BooleanType);
+    ISIN_CASE(UInt8Type);
+    ISIN_CASE(Int8Type);
+    ISIN_CASE(UInt16Type);
+    ISIN_CASE(Int16Type);
+    ISIN_CASE(UInt32Type);
+    ISIN_CASE(Int32Type);
+    ISIN_CASE(UInt64Type);
+    ISIN_CASE(Int64Type);
+    ISIN_CASE(FloatType);
+    ISIN_CASE(DoubleType);
+    ISIN_CASE(Date32Type);
+    ISIN_CASE(Date64Type);
+    ISIN_CASE(Time32Type);
+    ISIN_CASE(Time64Type);
+    ISIN_CASE(TimestampType);
+    ISIN_CASE(BinaryType);
+    ISIN_CASE(StringType);
+    ISIN_CASE(FixedSizeBinaryType);
+    ISIN_CASE(Decimal128Type);
+    default:
+      break;
+  }
+#undef ISIN_CASE
+
+  CHECK_IMPLEMENTED(kernel, "isin", type);
+  *out = std::move(kernel);
```

Comment:
I think we should build up the data structure here for the "right" input using InvokeUnaryKernel/InvokeHash linked below, and construct the kernel directly with the MemoTable